### PR TITLE
feature(zjow): add middleware for ape-x structure pipeline

### PR DESCRIPTION
**Pull request by zjowowen**
Add middleware priority_calculator for calculate priority in collector.
Add middleware PeriodicalModelExchanger for better control model send and receive process.

### CODE REVIEW

File: ding/data/buffer/middleware/priority.py [hunk 1]
```diff
@@ -54,7 +54,10 @@ def __init__(
 
     def push(self, chain: Callable, data: Any, meta: Optional[dict] = None, *args, **kwargs) -> BufferedData:
         if meta is None:
-            meta = {'priority': self.max_priority}
+            if 'priority' in data:
+                meta = {'priority': data['priority'].item()}
```
Comment (PaParaZz1) [inline ding/data/buffer/middleware/priority.py:58]: do item operation in original priority producer

File: ding/data/buffer/middleware/priority.py [hunk 2]
```diff
@@ -54,7 +54,10 @@ def __init__(
...
+            else:
```
Comment (PaParaZz1) [inline ding/data/buffer/middleware/priority.py:59]: add unittest for this new if-else branch

File: ding/framework/middleware/tests/test_distributer.py [hunk 3]
```diff
@@ -221,3 +221,54 @@ def pred(ctx):
...
+            task.use(pred)
+        task.run(8)
+
+
+@pytest.mark.tmp
+def test_periodical_model_exchanger():
+    Parallel.runner(n_parallel_workers=2, startup_interval=0)(periodical_model_exchanger_main)
+
+
+if __name__ == "__main__":
```
Comment (PaParaZz1) [inline ding/framework/middleware/tests/test_distributer.py:272]: unittest doesn't need this entry

File: ding/framework/middleware/functional/priority.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    """
+
+    if task.router.is_active and not task.has_role(task.role.COLLECTOR):
+        return task.void()
+
+    def _priority_calculator(ctx: "OnlineRLContext") -> None:
+
+        priority = func_for_priority_calculation(ctx.trajectories)
+        for i in range(len(priority)):
+            ctx.trajectories[i]['priority'] = torch.tensor(priority[i], dtype=torch.float32).unsqueeze(-1)
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/priority.py:23]: don't transform it to tensor

File: ding/framework/middleware/functional/priority.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        - func_for_priority_calculation (:obj:`Callable`): The function that calculates \
+            the priority of the collected data.
+    """
+
+    if task.router.is_active and not task.has_role(task.role.COLLECTOR):
+        return task.void()
+
+    def _priority_calculator(ctx: "OnlineRLContext") -> None:
+
+        priority = func_for_priority_calculation(ctx.trajectories)
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/priority.py:21]: priority_calculation_fn

File: ding/framework/middleware/functional/priority.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
+from typing import Callable
+import torch
+from ding.framework import task
+from ding.framework import OnlineRLContext
+
+
+def priority_calculator(func_for_priority_calculation: Callable) -> Callable:
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/priority.py:7]: add unittest for this file

File: ding/framework/middleware/distributer.py [hunk 7]
```diff
@@ -287,3 +288,140 @@ def _send_callback(self, storage: Storage):
...
+            - delay_toleration (:obj:`float`): Delay toleration of model exchange.
+            - stale_toleration (:obj:`int`): Stale toleration of model exchange.
+            - event_name (:obj:`str`): Event name of model exchange.
+            - model_loader (:obj:`ModelLoader`): Encode model in subprocess.
+        """
+        self._model = model
+        self._model_loader = model_loader
+        self._event_name = event_name
+        self._period = period
+        self.mode = mode
```
Comment (PaParaZz1) [inline ding/framework/middleware/distributer.py:321]: why not underline here

File: ding/framework/middleware/distributer.py [hunk 8]
```diff
@@ -287,3 +288,140 @@ def _send_callback(self, storage: Storage):
...
+        self._state_dict_cache: Optional[Union[object, Storage]] = None
+
+        if self.mode == "receive":
+            task.on(self._event_name, self._cache_state_dict)
+        if model_loader:
+            task.once("finish", lambda _: model_loader.shutdown())
+
+    def _cache_state_dict(self, msg: Dict[str, Any]):
+        # msg: Dict {'id':id,'model':state_dict: Union[object, Storage]}
+        print(f"node_id[{task.router.node_id}] get model msg")
```
Comment (PaParaZz1) [inline ding/framework/middleware/distributer.py:339]: use logging

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/data/buffer/middleware/priority.py (hunk 1 above)
Comment (zjowowen) [reply ding/data/buffer/middleware/priority.py]: fixed

File: ding/data/buffer/middleware/priority.py (hunk 2 above)
Comment (zjowowen) [reply ding/data/buffer/middleware/priority.py:59]: fixed

File: ding/framework/middleware/tests/test_distributer.py (hunk 3 above)
Comment (zjowowen) [reply ding/framework/middleware/tests/test_distributer.py]: fixed

File: ding/framework/middleware/functional/priority.py (hunk 4 above)
Comment (zjowowen) [reply ding/framework/middleware/functional/priority.py]: fixed

File: ding/framework/middleware/functional/priority.py (hunk 5 above)
Comment (zjowowen) [reply ding/framework/middleware/functional/priority.py]: fixed

File: ding/framework/middleware/distributer.py (hunk 7 above)
Comment (zjowowen) [reply ding/framework/middleware/distributer.py]: fixed

File: ding/framework/middleware/distributer.py (hunk 8 above)
Comment (zjowowen) [reply ding/framework/middleware/distributer.py]: fixed

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**